### PR TITLE
Add concat grad cinn

### PR DESCRIPTION
**Pull request by wangzhen38**
### PR types
Others

### PR changes
Others

### Describe
拆分concat算子的反向过程

### CODE REVIEW

File: paddle/fluid/prim/api/manual_prim/static_prim_api.cc [hunk 1]
```diff
@@ -120,6 +120,34 @@ Tensor full<DescTensor>(const IntArray& shape,
   op->InferShape(*block);
   return out;
 }
+
+template <>
+std::vector<Tensor> split<DescTensor>(const Tensor& x,
+                                      const IntArray& sections,
+                                      const Scalar& axis) {
+  // Grad infershape
```
Comment (xiaoguoguo626807) [inline paddle/fluid/prim/api/manual_prim/static_prim_api.cc]: describe it clearly or delete

File: python/paddle/fluid/tests/unittests/test_concat_op.py [hunk 2]
```diff
@@ -61,13 +63,13 @@ def test_check_output(self):
     def test_check_grad(self):
         if self.dtype == np.uint16:
             place = core.CUDAPlace(0)
-            self.check_grad_with_place(place, ['x0'], 'Out')
-            self.check_grad_with_place(place, ['x1'], 'Out')
-            self.check_grad_with_place(place, ['x2'], 'Out')
+            self.check_grad_with_place(place, ['x0'], 'Out', check_prim=False)
```
Comment (Charles-hit) [inline python/paddle/fluid/tests/unittests/test_concat_op.py]: 这儿为什么是False？

File: paddle/fluid/prim/api/manual_prim/static_prim_api.cc (hunk 1 above)
Comment (wangzhen38) [reply paddle/fluid/prim/api/manual_prim/static_prim_api.cc]: done

File: python/paddle/fluid/tests/unittests/test_concat_op.py (hunk 2 above)
Comment (wangzhen38) [reply python/paddle/fluid/tests/unittests/test_concat_op.py]: 已修改

File: paddle/fluid/operators/concat_op.cc [hunk 3]
```diff
@@ -153,6 +156,33 @@ class ConcatGradOpMaker : public framework::SingleGradOpMaker<T> {
...
+    std::vector<paddle::experimental::Tensor *> dx_ptr =
+        this->GetOutputPtr(input_grad_ptr);
+    std::vector<std::string> dx_name = this->GetOutputName(input_grad);
+
+    VLOG(6) << "Runing concat_grad composite func";
+    prim::concat_grad<prim::DescTensor>(input, out_grad, axis, dx_ptr);
+    this->RecoverOutputName(input_grad, dx_name);
+  }
+};
+
```
Comment (xiaoguoguo626807) [inline paddle/fluid/operators/concat_op.cc:193]: compositeOPMaker should be registered in REGISTER_OPERATOR(concat,

Comment (wangzhen38) [reply paddle/fluid/operators/concat_op.cc:193]: done

File: paddle/fluid/operators/concat_op.cc [hunk 4]
```diff
@@ -153,6 +156,33 @@ class ConcatGradOpMaker : public framework::SingleGradOpMaker<T> {
...
+    paddle::experimental::Tensor out_grad = this->GetSingleOutputGrad("Out");
+    std::vector<paddle::experimental::Tensor> input_grad =
+        this->GetMultiForwardInput("X");
+
+    std::vector<paddle::experimental::Tensor *> input_grad_ptr(
+        input_grad.size());
+    for (auto sub_tensor : input_grad) {
+      input_grad_ptr.push_back(&sub_tensor);
+    }
+    int axis = static_cast<int>(this->Attr<int>("axis"));
```
Comment (JiabinYang) [inline paddle/fluid/operators/concat_op.cc:177]: Maybe raise error when we used AxisTensor

Comment (JiabinYang) [reply paddle/fluid/operators/concat_op.cc:177]: Like what gather op did

**Review by JiabinYang — COMMENTED** [review]
some comments

File: paddle/fluid/operators/concat_op.cc (hunk 4 above)
Comment (wangzhen38) [reply paddle/fluid/operators/concat_op.cc:177]: done

**Review by JiabinYang — APPROVED** [review]
LGTM

**Review by seemingwang — APPROVED** [review]

**Pull request merged**